lib/object.c

```diff
@@ -188,44 +172,71 @@ tool_rc setup_primary(ESYS_CONTEXT *esys_ctx, ESYS_TR *parent) {
     return rc;
 }
 
+tool_rc tpm2_util_object_fetch_tpk(const char *objectstr,
+        TSSPRIVKEY_OBJ **tpk) {
+    tool_rc rc = tool_rc_success;
+    BIO *input_bio = BIO_new_file(objectstr, "rb");
+    if (!input_bio) {
+        LOG_ERR("Unable to read as BIO file");
+        rc = tool_rc_general_error;
+        goto ret;
+    }
+
+    *tpk = PEM_read_bio_TSSPRIVKEY_OBJ(input_bio, NULL, NULL, NULL);
+    if (*tpk == NULL) {
+        LOG_ERR("Unable to read PEM from provided BIO/file");
+        rc = tool_rc_general_error;
+        goto ret;
+    }
+
+ret:
+    if (input_bio) {
+        BIO_free(input_bio);
+    }
+    return rc;
+}
+
+tool_rc tpm2_util_object_fetch_priv_pub_from_tpk(TSSPRIVKEY_OBJ *tpk,
+        TPM2B_PUBLIC *pub, TPM2B_PRIVATE *priv) {
+
+    int pub_len = tpk->pubkey->length;
+    int priv_len = tpk->privkey->length;
+            
+    tool_rc rc = Tss2_MU_TPM2B_PUBLIC_Unmarshal(tpk->pubkey->data, pub_len,
+        NULL, pub);
+    if (rc != tool_rc_success) {
+        LOG_ERR("Error deserializing public portion of object");
+        goto out;
+    }
+
+    rc = Tss2_MU_TPM2B_PRIVATE_Unmarshal(tpk->privkey->data, priv_len,
+        NULL, priv);
+    if (rc != tool_rc_success) {
+        LOG_ERR("Error deserializing private portion of object");
+        goto out;
+    }
+
+out:
```

Comment:
A goto here is not really doing anything, like freeing data, so goto out could be replaced with return rc and then we don't need this label....